snuba/query/validation/__init__.py

```diff
@@ -17,5 +16,5 @@ class FunctionCallValidator(ABC):
     """
 
     @abstractmethod
-    def validate(self, parameters: Sequence[Expression], schema: ColumnSet) -> None:
+    def validate(self, parameters: Sequence[Expression], schema: Any) -> None:
```

Comment:
@fpacifici i think im unsure how to handle the circular import here because `Entity` imports the `FunctionCallValidator` for https://github.com/getsentry/snuba/blob/master/snuba/datasets/entity.py#L95

Reply:
If the circular import is because of typing, you can use the `typing.TYPE_CHECKING` value.

https://mypy.readthedocs.io/en/stable/runtime_troubles.html#import-cycles

That seems like it might make sense here, since the two classes are not really interconnected.

Reply:
I don't think you need to have any circular import. 
You can rely on the Entity class in the data sources module: https://github.com/getsentry/snuba/blob/master/snuba/query/data_source/simple.py#L32

That is present in the query, thus can easily be passed to a validator. The validator can fetch the full entity object from the entity key.

This way:
- the validators that are embedded in query processors like `HandledFunctionsProcessor` won't need an entity to be instantiated. They will pick it up in the process_query method from the query. The fact you were passing `self` to an object you instantiate statically seems a smell as it calls out a real circular dependency.
- There will not be any circular dependency between Entity (the dataset class) and the validator.
- The AllowedFunctionValidator can instead depend on the Entity as it should not be referenced by any Entity object.

The reasoning to have both Entity (in the dataset module) and a different Entity as a data source is summarized in this:
https://github.com/getsentry/snuba/blob/master/snuba/datasets/dataset.py#L36-L40

It is meant to allow the query processing to manipulate the query components (including the parts of the data model) without having to reference the data model definition classes (what descends from dataset) thus without ever creating circular dependencies. 
You will notice that Query, RequestSettings, DataSource(Entity/Table), etc. never depend on data model definition (Dataset, Entity, QueryProcessor, etc.) but the opposite is true: a QueryProcessor for example depends on Query.

So I think it would be ok to make `FunctionCallsValidator` depend on Entity (the data model one) as it is not referenced by any data model component. While `FunctionCallValidator` (singular) would not in its abstract interface depend on Entity.